core/ide/src/main/java/org/alice/ide/member/FunctionTabComposite.java

```diff
@@ -103,23 +105,45 @@ private List<MethodsSubComposite> getByReturnTypeSubComposites() {
     if (instanceFactory != null) {
       AbstractType<?, ?, ?> type = instanceFactory.getValueType();
       while (type != null) {
-        for (AbstractMethod method : type.getDeclaredMethods()) {
-          AbstractType<?, ?, ?> returnType = method.getReturnType();
-          if (returnType != JavaType.VOID_TYPE && isInclusionDesired(method)) {
-            List<AbstractMethod> methods = funcByType.computeIfAbsent(returnType, k -> Lists.newLinkedList());
-            methods.add(method);
+        if (type instanceof NamedUserType) {
+          NamedUserType namedUserType = (NamedUserType) type;
+
+          UserMethodsSubComposite userMethodsSubComposite = this.getUserMethodsSubComposite(namedUserType);
+
```

Comment:
this is long and repeat-y  maybe move lines 113-119/121-127 into a separate function?

Reply:
Repeaty, yes.
It bothers me that you needed to get the methods from an AbstractType in a different way for JavaType and NamedUserType, but I can see they do produce different results.

I would go as far as pushing 107-130 onto AbstractType and its subclasses. Have the type walk its own hierarchy to return the list of methods, then consider that list here, and above in Members.

Reply:
I'd originally taken that from getSubCompositesByCategory and put it into the other two getGetSubcomposites functions without changing it too much. I was considering splitting it off somehow to reduce repetition as well. Will see about moving it into AbstractType and subclasses.